### DISCUSSION TITLE
How to "connect" a generic type with a `z.ZodType` inferring to that type

### DISCUSSION THREAD

**Discussion by Svish** [colinhacks/zod#1310] [Q&A] [answered]
Not sure if the wording of the question is correct, but what I'm trying to do is to figure out how to type a function like this correctly:
```ts
function safeGet<T, S extends z.ZodType<T>>(key: string, s: S): z.infer<S> {
  try {
    const item = window.sessionStorage.getItem(key);
    return s.parse(item);
  } catch (e) {
    throw e;
  }
}

const x1 = safeGet('key', z.string());
// T is `unknown`, want it to be `string`

const x2 = safeGet('key', z.string().nullable());
// T is `unknown`, want it to be `string | null`

const y1 = safeGet('key', z.number());
// T is `unknown`, want it to be `number`

const y2 = safeGet('key', z.number().nullable());
// T is `unknown`, want it to be `number | null`
```
In all these 4 examples, the type `T` infers to `unknown`, and I don't understand how to get it to be what I want.

The docs on [Writing generic functions](https://zod.dev/?id=writing-generic-functions) mentions you should use `z.ZodTypeAny`, if I understand correctly, but if I switch to that I get an issue with unsafe any:
```ts
function safeGet<S extends z.ZodTypeAny>(key: string, s: S): z.infer<S> {
  try {
    const item = window.sessionStorage.getItem(key);
    return s.parse(item);
    // Unsafe return of an `any` typed value. eslint[@typescript-eslint/no-unsafe-return]
  } catch (e) {
    throw e;
  }
}
```

Obviously something is off, or "disconnected" somehow, but I just can't figure this one out. 😕

Anyone able to shed some light on how to type this properly?

**Comment by nullndr** [comment, marked as answer]
Shift `T`

```ts
function safeGet<S extends z.ZodType, T = z.infer<S>>(key: string, s: S): z.infer<S>
```

Check the [playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBALzgMyhEcBECIBNMBQByArgHYDGMwEZcAzgIbICmA4izADwDKcLADxgsyueogB0ALTwAVAJ5gWAGjiy4AXknAyrKLwB8BgBQBrFvIBcDGFB0BzVfWs8AlNYQSdew3ADeBHBwtvL+gUFwFLT08MDCGFoA7jq4EIkS9Cz09DRkPDDQjPYsEsUwAJLxZhauANzhQVCcJFB09BJgjFCZxnEsIHXhAL6RjDAUABZwxiyuYRHBE2iJ-PVBQwQbBFFkMXACAIyaDMzsnMYA5ObyF6qeMXZk9saugwD0b2pwwOIABg8OX5EHZ7AQAJmOTFYHBgl2ut0kAKeLwkZBIABt0YwAEbolgvd6fdQ-OD-WwOOAAHzgaMxQO20Xg8iOWihZ1hVwsCM8aJA2JYUAJ9Q+XxJv15-Kg9JBTIhrNOMLhXLuqJIfIFKNpWNx+NewqJ3z+EoFVJpGPRvyAA)

**Reply by nullndr** [reply]
@Svish does this answer your questions?

**Reply by Svish** [reply]
@nullndr Back at the office after the weekend today, and yeah, it does seem to help, for some reason. Not sure I understand why it helps to just switch the order like that, especially the type `T` isn't even used anywhere, like in your playground.

But, it does indeed seem to help, and adding an `extends z.infer<S>` made the version with fallback, which is what I actually have in my code, work as well:

```ts
function safeGet<S extends z.ZodType, F extends z.infer<S> = z.infer<S>>(key: string, s: S, fallback: F): z.infer<S> {
  try {
    const item = window.sessionStorage.getItem(key);
    if(item == null) return fallback;
    return s.parse(item);
  } catch (e) {
    return fallback;
  }
}
```

[playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBALzgMyhEcBECIBNMBQByArgHYDGMwEZcAzgIbICmA4izADwDKcLADxgsyueogB0ALTwAVAJ5gWAGjgAxfkJFjJwMqyi8AfHAC8u-S0M8jRgBQBrFvIBcDGFD0BzVfTc9VZEYAG2CAI0YKBzc1AEo3BAk9A2M4AG8CODgPeXTMrLgKWnp4YGEMcwB3PVwISol6Fnp6GjIeGGhGLxYJbpgASXLHZ1iAbnys4GQ7MpYK8zISUNi4KE4SKDog0Iio8YLV9c2GCTBGKEaZ8rH8gF9CxhgKAAs4OxYVjIO1mA2tkPCkQc+zgtwIYIIRTIJTgAgAjGYGMx2Jw7AByJzyNGqRIlTxkLx2WKqNFom4AenJcFkcGA4gABnjvPSiFCYQIAEyIpisDgwdGY7GSJkEokSRahRhhYIsImqCXBClUml0uCMjzeOAAHzgCpZkOK8HkCPMPJR-IxziFiUWIDCVjlcAADErqbSGbb7VB9WyjVzTci+QKrTjxSQ7Q7YmHJdLZcTdUtFeNKW7VfTPVZtQnQj7DYgTUjeajLVjQxmoGLzlBGPJHQBtAC6rpVHvDXqzeqAA)

Don't really understand why though. Guess I'll just accept that it works and move on, haha 😂🤷‍♂️

**Reply by nullndr** [reply]
In your function `T` is evaluated before `S`, but TS can not determine it's type because there is actually nothing that use it, so TS assign `unknown` to T.

In my example, `S` is used by `s`, which is a `z.ZodType`, then `T` is evaluated as `z.infer<S>`, and TS can assign to him the right type.

**Reply by Svish** [reply]
Aah, ok, so generic types are actually processed one-by-one, in order. Think I've seen them as linked only when they actually refer to each other, and that they then were processed depending on what type depends on what other type.

Makes more sense now, thanks!
